source/custom/visa_service.custom.cpp

```diff
@@ -320,28 +320,25 @@ ::grpc::Status VisaService::Read(::grpc::ServerContext* context, const ReadReque
   if (context->IsCancelled()) {
     return ::grpc::Status::CANCELLED;
   }
-#if 1
-  return ::grpc::Status(grpc::StatusCode::DO_NOT_USE, "Custom code not implemented yet");
-#else
   try {
     auto vi_grpc_session = request->vi();
     ViSession vi = session_repository_->access_session(vi_grpc_session.name());
     ViUInt32 count = request->count();
-    std::string buffer(return_count, '\0');
+    std::string buffer(256 - 1, '\0');
```

Comment:
The size of the buffer needs to be `count` bytes. This is not a string and doesn't need to be 0 initialized. Read operations can contain purely binary data. I'd suggest this:
`std::vector<ViByte> buffer(count);`

Reply:
I think this would also be one of the very few places where you should add a catch block for `std::bad_alloc&`. In that case you should return the gRPC unknown code with NI error code `VI_ERROR_ALLOC`.

Reply:
std::vector<ViBtye> buffer(count) is still allocated on the stack, I am not sure why it needs to check for `std::bad_alloc`. Does the `Read` in VISA throw? Since it returns an error, and it is a C API, it probably doesn't throw, right?

Reply:
That's not quite correct. The *vector* object is on the stack, yes, but that's only 24 bytes (3 pointers to heap locations).

Reply:
It turned out in the binding file `visa.pb.h`, it defines the read buffer as `::PROTOBUF_NAMESPACE_ID::internal::ArenaStringPtr buffer_;` which actually is a String type. I think that's why it generates the method with the std::string for the buffer, and we have to stick with it.

Reply:
Saying the proto type is `bytes` is technically correct. We must be extra cautious to not use "strlen" or related methods anywhere on this. In particular, this points out an important test case for a buffer such as `A\x00B\x00C` and make sure all 5 bytes are properly included.

Reply:
I'm generating the client side and realized I have to be careful because of this issue as well. It's a C/C++ binding limitation, not a protobuf limitation.